compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -202,8 +191,21 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         operand: OperandRef<'tcx, Bx::Value>,
         cast: TyAndLayout<'tcx>,
     ) -> Option<OperandValue<Bx::Value>> {
-        // Callers already checked that the layout sizes match
-        debug_assert_eq!(operand.layout.size, cast.size);
+        // Check for transmutes that are always UB.
+        if operand.layout.size != cast.size
+            || operand.layout.abi.is_uninhabited()
+            || cast.abi.is_uninhabited()
+        {
+            if !operand.layout.abi.is_uninhabited() {
+                // Since this is known statically and the input could have existed
+                // without already having hit UB, might as well trap for it.
+                bx.abort();
+            }
```

Comment:
No, for the `abort` block I really do mean the input could have existed.

Basically, if we're transmuting `!` → `_` then there's no need to trap, since to get here they already did something UB.  Whereas for `()` → `!` this is also UB, but it's (in a sense) the "first" UB, so I figured it'd be worth trapping for, since it must always be dead code and thus if it's actually in the final binary detecting it seems plausibly helpful.

But then for all three UB cases -- size mismatch, impossible input, and impossible output -- I just return poison instead of bothering with figuring out bitcasts or whatever.

Reply:
Oh actually, that clarification made me realize that I was reading this block totally wrong and that suggested refactor makes no sense, lol.